src/Mvc/src/Microsoft.AspNetCore.Mvc.ViewFeatures/ModelExpressionProvider.cs

```diff
@@ -2,43 +2,61 @@
-        private readonly ExpressionTextCache _expressionTextCache;
+        private readonly ConcurrentDictionary<LambdaExpression, string> _expressionTextCache;
 
         /// <summary>
         /// Creates a  new <see cref="ModelExpressionProvider"/>.
         /// </summary>
         /// <param name="modelMetadataProvider">The <see cref="IModelMetadataProvider"/>.</param>
-        /// <param name="expressionTextCache">The <see cref="ExpressionTextCache"/>.</param>
-        public ModelExpressionProvider(
-            IModelMetadataProvider modelMetadataProvider,
-            ExpressionTextCache expressionTextCache)
+        public ModelExpressionProvider(IModelMetadataProvider modelMetadataProvider)
         {
             if (modelMetadataProvider == null)
             {
                 throw new ArgumentNullException(nameof(modelMetadataProvider));
             }
 
-            if (expressionTextCache == null)
+            _modelMetadataProvider = modelMetadataProvider;
+            _expressionTextCache = new ConcurrentDictionary<LambdaExpression, string>(LambdaExpressionComparer.Instance);
+        }
+
+        /// <summary>
+        /// Gets the name for <paramref name="expression"/>.
+        /// </summary>
+        /// <typeparam name="TModel">The model type.</typeparam>
+        /// <typeparam name="TValue">The type of the <paramref name="expression"/> result.</typeparam>
+        /// <param name="expression">The expression.</param>
+        /// <returns>The expression name.</returns>
+        public string GetExpressionText<TModel, TValue>(Expression<Func<TModel, TValue>> expression)
+        {
+            if (expression == null)
             {
-                throw new ArgumentNullException(nameof(expressionTextCache));
+                throw new ArgumentNullException(nameof(expression));
             }
 
-            _modelMetadataProvider = modelMetadataProvider;
-            _expressionTextCache = expressionTextCache;
+            return ExpressionHelper.GetExpressionText(expression, _expressionTextCache);
         }
 
-        /// <inheritdoc />
+        /// <summary>
+        /// Returns a <see cref="ModelExpression"/> instance describing the given <paramref name="expression"/>.
+        /// </summary>
+        /// <typeparam name="TModel">The type of the <paramref name="viewData"/>'s <see cref="ViewDataDictionary{T}.Model"/>.</typeparam>
+        /// <typeparam name="TValue">The type of the <paramref name="expression"/> result.</typeparam>
+        /// <param name="viewData">The <see cref="ViewDataDictionary{TModel}"/> containing the <see cref="ViewDataDictionary{T}.Model"/>
+        /// against which <paramref name="expression"/> is evaluated. </param>
+        /// <param name="expression">An expression to be evaluated against the current model.</param>
+        /// <returns>A new <see cref="ModelExpression"/> instance describing the given <paramref name="expression"/>.</returns>
```

Comment:
Why not use the comments from `IModelExpressionProvider`? That is, what are you improving here -- if anything?